### PR TITLE
Adjust for non-compliant parsers

### PR DESCRIPTION
**Pull request by jacogr**

### CODE REVIEW

**Comment by paulmillr** [comment]
Can you run benchmarks before/after the change?

**Comment by jacogr** [comment]
Will do.

**Comment by jacogr** [comment]
`npm run bench` as per `main` (original) -

```
npm run bench      

> @noble/hashes@0.4.1 bench /Users/jacogr/Projects/polkadotjs/noble-hashes
> node test/benchmark/index.js

-------
Benchmarking
==== SHA256 ====
SHA256 32B node x 396,825 ops/sec @ 2μs/op
SHA256 32B hash-wasm x 1,001,001 ops/sec @ 999ns/op
SHA256 32B crypto-browserify x 567,536 ops/sec @ 1μs/op
SHA256 32B stablelib x 536,768 ops/sec @ 1μs/op
SHA256 32B fast-sha256 x 567,536 ops/sec @ 1μs/op
SHA256 32B noble x 759,878 ops/sec @ 1μs/op

==== SHA384 ====
SHA384 32B node x 419,639 ops/sec @ 2μs/op
SHA384 32B crypto-browserify x 202,634 ops/sec @ 4μs/op
SHA384 32B stablelib x 288,184 ops/sec @ 3μs/op
SHA384 32B noble x 345,901 ops/sec @ 2μs/op

==== SHA512 ====
SHA512 32B node x 418,410 ops/sec @ 2μs/op
SHA512 32B hash-wasm x 692,520 ops/sec @ 1μs/op
SHA512 32B crypto-browserify x 213,037 ops/sec @ 4μs/op
SHA512 32B stablelib x 294,637 ops/sec @ 3μs/op
SHA512 32B noble x 354,735 ops/sec @ 2μs/op

==== SHA3-256, keccak256, shake256 ====
SHA3-256, keccak256, shake256 32B node x 376,647 ops/sec @ 2μs/op
SHA3-256, keccak256, shake256 32B hash-wasm x 707,714 ops/sec @ 1μs/op
SHA3-256, keccak256, shake256 32B stablelib x 298,775 ops/sec @ 3μs/op
SHA3-256, keccak256, shake256 32B js-sha3 x 375,516 ops/sec @ 2μs/op
SHA3-256, keccak256, shake256 32B noble x 117,370 ops/sec @ 8μs/op

==== Kangaroo12 ====
Kangaroo12 32B noble x 183,318 ops/sec @ 5μs/op

==== Marsupilami14 ====
Marsupilami14 32B noble x 166,694 ops/sec @ 5μs/op

==== BLAKE2b ====
BLAKE2b 32B node x 415,627 ops/sec @ 2μs/op
BLAKE2b 32B hash-wasm x 684,462 ops/sec @ 1μs/op
BLAKE2b 32B stablelib x 211,327 ops/sec @ 4μs/op
BLAKE2b 32B noble x 225,886 ops/sec @ 4μs/op

==== BLAKE2s ====
BLAKE2s 32B node x 414,078 ops/sec @ 2μs/op
BLAKE2s 32B hash-wasm x 1,013,171 ops/sec @ 987ns/op
BLAKE2s 32B stablelib x 1,072,961 ops/sec @ 932ns/op
BLAKE2s 32B noble x 374,251 ops/sec @ 2μs/op

==== BLAKE3 ====
BLAKE3 32B hash-wasm x 1,228,501 ops/sec @ 814ns/op
BLAKE3 32B noble x 384,911 ops/sec @ 2μs/op

==== RIPEMD160 ====
RIPEMD160 32B node x 425,713 ops/sec @ 2μs/op
RIPEMD160 32B crypto-browserify x 484,496 ops/sec @ 2μs/op
RIPEMD160 32B noble x 667,556 ops/sec @ 1μs/op

==== HMAC-SHA256 ====
HMAC-SHA256 32B node x 293,685 ops/sec @ 3μs/op
HMAC-SHA256 32B crypto-browserify x 169,491 ops/sec @ 5μs/op
HMAC-SHA256 32B stablelib x 212,494 ops/sec @ 4μs/op
HMAC-SHA256 32B noble x 220,264 ops/sec @ 4μs/op

RAM: rss=454.5mb heap=27.9mb used=7.7mb ext=2.5mb arr=0.4mb
-------
Benchmarking
==== PBKDF2-HMAC-SHA256 ====
PBKDF2-HMAC-SHA256 16384 node x 126 ops/sec @ 7ms/op
PBKDF2-HMAC-SHA256 16384 stablelib x 41 ops/sec @ 23ms/op
PBKDF2-HMAC-SHA256 16384 noble x 32 ops/sec @ 30ms/op
PBKDF2-HMAC-SHA256 16384 noble (async) x 30 ops/sec @ 32ms/op

PBKDF2-HMAC-SHA256 65536 node x 32 ops/sec @ 30ms/op
PBKDF2-HMAC-SHA256 65536 stablelib x 11 ops/sec @ 84ms/op
PBKDF2-HMAC-SHA256 65536 noble x 8 ops/sec @ 112ms/op
PBKDF2-HMAC-SHA256 65536 noble (async) x 8 ops/sec @ 121ms/op

PBKDF2-HMAC-SHA256 262144 node x 8 ops/sec @ 124ms/op
PBKDF2-HMAC-SHA256 262144 stablelib x 3 ops/sec @ 332ms/op
PBKDF2-HMAC-SHA256 262144 noble x 2 ops/sec @ 443ms/op
PBKDF2-HMAC-SHA256 262144 noble (async) x 2 ops/sec @ 486ms/op

==== PBKDF2-HMAC-SHA512 ====
PBKDF2-HMAC-SHA512 16384 node x 110 ops/sec @ 9ms/op
PBKDF2-HMAC-SHA512 16384 hash-wasm x 18 ops/sec @ 54ms/op
PBKDF2-HMAC-SHA512 16384 stablelib x 11 ops/sec @ 87ms/op
PBKDF2-HMAC-SHA512 16384 noble x 13 ops/sec @ 75ms/op
PBKDF2-HMAC-SHA512 16384 noble (async) x 13 ops/sec @ 76ms/op

PBKDF2-HMAC-SHA512 65536 node x 27 ops/sec @ 35ms/op
PBKDF2-HMAC-SHA512 65536 hash-wasm x 5 ops/sec @ 199ms/op
PBKDF2-HMAC-SHA512 65536 stablelib x 2 ops/sec @ 340ms/op
PBKDF2-HMAC-SHA512 65536 noble x 3 ops/sec @ 290ms/op
PBKDF2-HMAC-SHA512 65536 noble (async) x 3 ops/sec @ 304ms/op

PBKDF2-HMAC-SHA512 262144 node x 6 ops/sec @ 144ms/op
PBKDF2-HMAC-SHA512 262144 hash-wasm x 1 ops/sec @ 796ms/op
PBKDF2-HMAC-SHA512 262144 stablelib x 0 ops/sec @ 1369ms/op
PBKDF2-HMAC-SHA512 262144 noble x 0 ops/sec @ 1161ms/op
PBKDF2-HMAC-SHA512 262144 noble (async) x 0 ops/sec @ 1220ms/op

==== Scrypt r: 8, p: 1, n: ====
Scrypt r: 8, p: 1, n: 16384 node x 24 ops/sec @ 40ms/op
Scrypt r: 8, p: 1, n: 16384 scrypt-async x 14 ops/sec @ 67ms/op
Scrypt r: 8, p: 1, n: 16384 scrypt-js x 7 ops/sec @ 137ms/op
Scrypt r: 8, p: 1, n: 16384 stablelib x 17 ops/sec @ 55ms/op
Scrypt r: 8, p: 1, n: 16384 noble x 16 ops/sec @ 60ms/op

Scrypt r: 8, p: 1, n: 65536 node x 5 ops/sec @ 169ms/op
Scrypt r: 8, p: 1, n: 65536 scrypt-async x 3 ops/sec @ 263ms/op
Scrypt r: 8, p: 1, n: 65536 scrypt-js x 1 ops/sec @ 570ms/op
Scrypt r: 8, p: 1, n: 65536 stablelib x 4 ops/sec @ 205ms/op
Scrypt r: 8, p: 1, n: 65536 noble x 4 ops/sec @ 242ms/op

Scrypt r: 8, p: 1, n: 262144 node x 1 ops/sec @ 783ms/op
Scrypt r: 8, p: 1, n: 262144 scrypt-async x 0 ops/sec @ 1091ms/op
Scrypt r: 8, p: 1, n: 262144 scrypt-js x 0 ops/sec @ 2405ms/op
Scrypt r: 8, p: 1, n: 262144 stablelib x 1 ops/sec @ 857ms/op
Scrypt r: 8, p: 1, n: 262144 noble x 0 ops/sec @ 1006ms/op

==== Scrypt Async r: 8, p: 1, n: ====
Scrypt Async r: 8, p: 1, n: 16384 node x 24 ops/sec @ 41ms/op
Scrypt Async r: 8, p: 1, n: 16384 hash-wasm x 22 ops/sec @ 44ms/op
Scrypt Async r: 8, p: 1, n: 16384 scrypt-async x 12 ops/sec @ 81ms/op
Scrypt Async r: 8, p: 1, n: 16384 scrypt-js x 3 ops/sec @ 255ms/op
Scrypt Async r: 8, p: 1, n: 16384 stablelib x 14 ops/sec @ 71ms/op
Scrypt Async r: 8, p: 1, n: 16384 noble x 13 ops/sec @ 75ms/op

Scrypt Async r: 8, p: 1, n: 65536 node x 5 ops/sec @ 168ms/op
Scrypt Async r: 8, p: 1, n: 65536 hash-wasm x 5 ops/sec @ 176ms/op
Scrypt Async r: 8, p: 1, n: 65536 scrypt-async x 3 ops/sec @ 321ms/op
Scrypt Async r: 8, p: 1, n: 65536 scrypt-js x 0 ops/sec @ 1004ms/op
Scrypt Async r: 8, p: 1, n: 65536 stablelib x 3 ops/sec @ 275ms/op
Scrypt Async r: 8, p: 1, n: 65536 noble x 3 ops/sec @ 290ms/op

Scrypt Async r: 8, p: 1, n: 262144 node x 1 ops/sec @ 745ms/op
Scrypt Async r: 8, p: 1, n: 262144 hash-wasm x 1 ops/sec @ 733ms/op
Scrypt Async r: 8, p: 1, n: 262144 scrypt-async x 0 ops/sec @ 1378ms/op
Scrypt Async r: 8, p: 1, n: 262144 scrypt-js x 0 ops/sec @ 4071ms/op
Scrypt Async r: 8, p: 1, n: 262144 stablelib x 0 ops/sec @ 1132ms/op
Scrypt Async r: 8, p: 1, n: 262144 noble x 0 ops/sec @ 1240ms/op

RAM: rss=1277.0mb heap=10.0mb used=5.8mb ext=270.4mb arr=268.5mb
```

File: src/sha3.ts [hunk 1]
```diff
@@ -10,19 +11,26 @@ import {
...
   // Rotational
   SHA3_ROTL.push((((round + 1) * (round + 2)) / 2) % 64);
   // Iota
-  let t = 0n;
+  let t = _0n;
   for (let j = 0; j < 7; j++) {
-    R = ((R << 1n) ^ ((R >> 7n) * 0x71n)) % 256n;
-    if (R & 2n) t ^= 1n << ((1n << BigInt(j)) - 1n);
+    R = ((R << _1n) ^ ((R >> _7n) * _0x71n)) % _256n;
+    if (R & _2n) t ^= _1n << ((_1n << BigInt(j)) - _1n);
```
Comment (jacogr) [inline src/sha3.ts:33]:
```suggestion
    if (R & _2n) t ^= _1n << ((_1n << _n(j)) - _1n);
```

Also a top-level expectation of `BigInt` being available.

**Pull request closed without merging**